### DISCUSSION TITLE
Understanding email link in rodauth mailer

### DISCUSSION THREAD

**Discussion by pozzos** [janko/rodauth-rails#177] [Q&A]
Hello!

I'm trying to customise the format of the links sent by rodauth's mailer, and there are a few lines I can't find documentation on. Since I have a JSON API, I need to redirect all these links to custom pages I have made that do AJAX calls.

For the account verification, modification seems easy (simply replace rodauth.verify_account_url with a custom function)

however for the others, there is an "email_link" function that seems to seems to set instance variables on rodauth. How does this work, and how could I go about modifying the format?

Apologies if this is an obvious question.

Best regards,
Sebastian

**Comment by janko** [comment]
Hi, it's not an obvious question at all 😉 The Action Mailer design is still something I'm iterating on, since there is no official recommendation in Rodauth itself on how best to background emails.

I modified the [`RodauthMailer`](https://github.com/janko/rodauth-demo-rails/blob/master/app/mailers/rodauth_mailer.rb) class in the demo app to keep the Rodauth instance. Rodauth doesn't seem to currently provide an easy way to get the email token, so you could create a wrapper method:

```rb
# app/misc/rodauth_main.rb
class RodauthMain < Rodauth::Rails::Auth
  configure do
    # ...
  end

  def verify_account_email_token
    token_param_value(verify_account_key_value)
  end
end
```

In your `RodauthMailer`, you could then build the frontend URL:

```rb
class RodauthMailer < ApplicationMailer
  def verify_account(name, account_id, key)
    @rodauth = rodauth(name, account_id) { @verify_account_key_value = key }
    @email_link = "https://frontend.example.com/verify-account?token=#{@rodauth.verify_account_email_token}"
    # ...
  end
end
```

Let me know if this works 😉 I think it would be useful to make the `token_param_value` Rodauth method public, I'll propose that to Jeremy.

**Reply by pozzos** [reply]
Dear Janko,

Thanks for such a detailed answer! I will try that and let you know how it goes.

Best regards,
Sebastian

**Reply by crespire** [reply]
The code provided for the `mailer` did not initially work. I changed it to this snippet which worked. I think it has to do with trying to pass in the account_id into the `rodauth` private method, which only takes name.

```rb
def verify_account(name, account_id, key)
  @rodauth = rodauth(name)
  @rodauth.instance_variable_set(:@account, { id: account_id })
  @rodauth.instance_variable_set(:@verify_account_key_value, key)
  @email_link = "http://localhost:3001/verify-account/#{@rodauth.verify_account_email_token}"
    # rest of code
end
```

**Reply by janko** [reply]
@crespire Yes, that code I posted works on a modified `RodauthMailer` that I recently pushed to the Rails demo repository.

**Comment by crespire** [comment]
This was helpful. I suppose if we wanted to verify other elements of the account, we would have to define the appropriate instance variable on `@rodauth`?

For example, if a user changed their login, then we'd have to define `verify_login_change_key_value` and use that to link to the front end. Similarly, it looks like for "Forgot Password" functionality, the variable would be `get_reset_password_key(id)` that we'd have to put in the link to the front end.

Just want to make sure I understand, thanks!

**Reply by janko** [reply]
Yes, this is what the [mailer in the demo Rails app](https://github.com/janko/rodauth-demo-rails/blob/33a1e030dc223031bc4b93812129654d3a1cb3a0/app/mailers/rodauth_mailer.rb) now does, it sets the corresponding `@*_key_value` instance variable, and uses the `*_email_link` method. To generate different links, you could override the corresponding `*_email_link` configuration method.

In the mailer, you don't need to retrieve raw tokens, because they're already passed in the mailer's arguments.

**Reply by crespire** [reply]
Perfect. I may try to update my mailer to mirror the sample app then. I've just been using the setup as is from when I ran the install command.

Thanks again! I am happy that you are active, it really gives me confidence to continue to stick with this gem over something like devise, which seems to have fallen into unmaintained status.

**Reply by crespire** [reply]
I wasn't able to get it working after switching the mailer to the new version. I kept getting scope.session no method on nil stemming from the `@account.rails_account`

I tried to update my Application controller as I thought maybe it was the lack of helper method to match the sample app, but I see it was not generated using the `--api` flag, and uses `ActionController::Base` rather than `ActionController::API`

I switched back to the version I posted, as it was working. Happy to share a repo for you to take a look if you'd like.

**Reply by janko** [reply]
Oh, I just found out this mailer setup doesn't work with latest rodauth-rails version. I just pushed a fix and released a new version, so that should be working now 🤞🏻

**Reply by pboling** [reply]
> To generate different links, you could override the corresponding *_email_link configuration method.

I'm want to do exactly that, but I don't know how.  The problem is I can't call `super`, and I can't find the definition of the `*_link` methods anywhere; I assume their are defined dynamically?

```
    verify_account_email_link do
      byebug
      super
    end
```

I am trying to figure out how to build my alternate version of a link (with a path key for _path-based_ multi-tenant support).

Note: calling `super` gives:
```
#<RuntimeError: implicit argument passing of super from method defined by define_method() is not supported. Specify all arguments explicitly.>
```
which is the dead giveaway for metaprogramming 😆

**Reply by pboling** [reply]
I figured it out!

```
    verify_account_email_link do
      segs = verify_account_url(key: @verify_account_key_value).split('/')
      segs.insert(-2, request.env[:path_key])
      segs.join('/')
    end
```

And I end up with a URL which in testing env looks like: `:///multi/tenant/verify-account?key=3Aqg_sQut52kW1NvUtrm2dpkLMvdjbTtKs5JVKrLOBE`, which is exactly what I am looking for!

**Reply by janko** [reply]
Yeah, there is metaprogramming involved. The verify_account feature defines a default `verify_account_email_link` method on the Rodauth object. The `verify_account_email_link` *configuration* method exists to override that method, and that uses `define_method` internally. That configuration method is basically implemented like this:

```rb
def verify_account_email_link(&block)
  @auth.define_method(:verify_account_email_link, &block)
end
```

Given that the block passed to `verify_account_email_link` is forwarded to `define_method`, Ruby requires you to call super with parentheses. If there are no arguments, then you need to specify empty parentheses:

```rb
verify_account_email_link do
  byebug
  super()
end
```

**Reply by pboling** [reply]
Thanks!  I have all that working, however, now when the test (I'm setting this up inside teh gem's test suite) is failing after visiting the link generated link.

I added a prefixed method to the `test_helper.rb`, to assist on the test side:
```
  def prefixed(path_key: nil, prefix: nil)
    [prefix, path_key].compact.select {|x| !x.empty? }.join("/")
  end
```

I am here now:
```
   67:     if verify
   68:       email = ActionMailer::Base.deliveries.last
   69:       verify_account_link = email.body.to_s[%r{#{prefixed(path_key: path_key, prefix: prefix)}/verify-account\S+}]
   70:       byebug
=> 71:       visit verify_account_link
   72:       click_on "Verify Account"
   73:     end
```
and I have this:
```
(byebug) verify_account_link
"/multi/tenant/kiwi/verify-account?key=Tkva-IFqky8i-73klUUCAiX5MVucfudvYzvS8b36e5c"
(byebug) visit verify_account_link
*** ActionController::RoutingError Exception: No route matches [GET] "/multi/tenant/login"

nil
```

I can't figure out where that login link is being generated, but regardless it is obvious that I must also override all the `*_path` methods to include the path key... and I'm not sure how to do that.

**Reply by janko** [reply]
It's difficult to tell what's going on with just bits and pieces. It's probably best to continue in https://github.com/janko/rodauth-rails/discussions/257, which is still pending my reply. I'm currently trying to solve a similar problem – generating localized routes (https://github.com/jeremyevans/rodauth/discussions/380) – so I wanted to use knowledge from that to make better suggestions.

**Comment by crespire** [comment]
I wanted to check back in. I have updated my application to use the reference mailer implementation (seen in the rails example app) as I ended up needing to modify more than just one mail link (initially, I was only concerned about account verify, but now I am working on password reset and login change).

Thank you so much for your conversations here, I updated the implementation without any issues on the latest `rodauth-rails` gem version!
